File: 025. Reverse Nodes in k-Group/Java/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode fakeHead = new ListNode(0);
        fakeHead.next = head;
        ListNode tail = fakeHead;
        ListNode current = fakeHead;
        while (true) {
            int count = k;
            while (count != 0 && tail != null) {
                tail = tail.next;
                count--;
            }
            if (tail == null) {
                break;
            }
            ListNode tempHead = current.next;
            while (current.next != tail) {
                ListNode node = current.next;
                current.next = node.next;
                node.next = tail.next;
                tail.next = node;
            }
            current = tempHead;
            tail = tempHead;
        }
        return fakeHead.next;
    }
}

File: 025. Reverse Nodes in k-Group/Python/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        fake_head = ListNode(0)
        fake_head.next = head
        tail = fake_head
        current = fake_head
        while True:
            count = k
            while count and tail:
                tail = tail.next
                count -= 1
            if not tail:
                break
            temp_head = current.next
            while current.next != tail:
                node = current.next
                current.next = node.next
                node.next = tail.next
                tail.next = node
            current = temp_head
            tail = temp_head
        return fake_head.next
